kafka/consume.go

```diff
@@ -0,0 +1,217 @@
+package kafka
+
+import (
+	"context"
+	"io"
+	"sync"
+	"time"
+
+	kafkalib "github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// ErrSeekTimedOut is the error returned when a consumer timed out during Seek.
+var ErrSeekTimedOut = errors.New("Kafka Seek timed out. Please try again.")
+
+// Consumer reads messages from Kafka.
+type Consumer interface {
+	io.Closer
+
+	// FetchMessage fetches one message, if there is any available at the current offset.
+	FetchMessage(ctx context.Context) (*kafkalib.Message, error)
+
+	// CommitMessage commits the offset of a given message.
+	CommitMessage(msg *kafkalib.Message) error
+}
+
+// OffsetAwareConsumer is a Consumer that can reset its offset.
+type OffsetAwareConsumer interface {
+	Consumer
+
+	// SetInitialOffset resets the current offset to the given one.
+	// Used for setting the initial offset a consumer should start consuming from.
+	// Should be called before start consuming messages.
+	SetInitialOffset(offset int64) error
+
+	// Seek seeks the assigned topic partitions to the given offset.
+	// Seek() may only be used for partitions already being consumed.
+	Seek(offset int64, timeout time.Duration) error
+}
+
+// ConfluentConsumer implements Consumer interface.
+type ConfluentConsumer struct {
+	c    *kafkalib.Consumer
+	conf Config
+	log  logrus.FieldLogger
+
+	rebalanceHandler      func(c *kafkalib.Consumer, ev kafkalib.Event) error // Only set when an initial offset should be set
+	rebalanceHandlerMutex sync.Mutex
+	subscribeOnce         sync.Once
+}
+
+// NewConsumer creates a ConfluentConsumer based on config.
+func NewConsumer(log logrus.FieldLogger, conf Config, opts ...ConfigOpt) (*ConfluentConsumer, error) {
+	// See Reference at https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+	kafkaConf := conf.baseKafkaConfig()
+	_ = kafkaConf.SetKey("enable.auto.offset.store", false) // manually StoreOffset after processing a message. Otherwise races may happen.)
+
+	// In case we try to assign an offset out of range (greater than log-end-offset), consumer will use start consuming from offset zero.
+	_ = kafkaConf.SetKey("auto.offset.reset", "earliest")
+
+	conf.Consumer.Apply(kafkaConf)
+	for _, opt := range opts {
+		opt(kafkaConf)
+	}
```

Comment:
should we flip these? This makes it soe that you can't override things like `enable.auto.offset.store`.

Reply:
Not sure if I understand correctly. You can actually override the defaults by just setting an option. As it runs after applying the default option, you can override it.

Reply:
Oh ~ wait. I read this wrong. I was thinking you were applying to something else. Nevermind. :)